linked_list/linked_list.py

```diff
@@ -2,78 +2,141 @@
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        counter = 0
+        current = self.head
+        while (current != None):
+            counter += 1
+            current = current.next
+        return counter
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current = self.head
+        i = 0
+        while (current != None):
+            if i == index:
+                return current.value
+            current = current.next
+            i += 1
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current = self.head
+        if self.head == None:
+            return None
+        while current.next!= None:
+            current = current.next
+        return current.value 
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if self.head == None:
+            self.add_first(value)
+        else:
+            current = self.head
+            while current.next != None:
+                current = current.next
+            current.next = Node(value)
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+    
     def find_max(self):
-        pass
+        current = self.head
+
+        if self.head == None:
+            return None
+        max = current.value
+        while (current != None):
+            if current.value > max:
+                max = current.value
+            current = current.next
+
+        return max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity:O(1)
     def delete(self, value):
-        pass
+        current = self.head
+        if self.head == None:
+            return None
+        if self.head.value == value:
+            self.head = self.head.next
+        else:
+            while current.next.value != value:
+                current = current.next
+            current.next = current.next.next

```

Comment:
You also need to adjust the `previous` links to maintain the doubly linked list.